core/src/connection/pool.rs

```diff
@@ -0,0 +1,907 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::{
+    Executor,
+    ConnectedPoint,
+    PeerId,
+    connection::{
+        self,
+        ConnectionId,
+        ConnectionLimit,
+        IncomingInfo,
+        OutgoingInfo,
+        Connection,
+        Substream,
+        Connected,
+        ConnectionError,
+        ConnectionHandler,
+        IntoConnectionHandler,
+        ConnectionInfo,
+        manager::{self, Manager},
+    },
+    muxing::StreamMuxer,
+};
+use either::Either;
+use fnv::FnvHashMap;
+use futures::prelude::*;
+use smallvec::SmallVec;
+use std::{error, fmt, hash::Hash, task::Context, task::Poll};
+
+/// A connection `Pool` manages a set of connections for each peer.
+pub struct Pool<TInEvent, TOutEvent, THandler, TTransErr, THandlerErr, TConnInfo = PeerId, TPeerId = PeerId> {
+    local_id: TPeerId,
+
+    /// The configuration of the pool.
+    limits: PoolLimits,
+
+    /// The connection manager that handles the connection I/O for both
+    /// established and pending connections.
+    ///
+    /// For every established connection there is a corresponding entry in `connected`.
```

Comment:
Not sure, but shouldn't it be:

```suggestion
    /// For every established connection there is a corresponding entry in `established`.
```

Reply:
Thanks!